torch/distributed/_tools/mod_tracker.py

```diff
@@ -0,0 +1,231 @@
+import warnings
+import weakref
+from typing import Callable, Optional, Set
+
+import torch
+from torch.autograd.graph import register_multi_grad_hook
+from torch.nn.modules.module import (
+    register_module_forward_hook,
+    register_module_forward_pre_hook,
+)
+from torch.utils._pytree import tree_flatten
+
+
+__all__ = ["ModTracker"]
+
+
+class ModTracker:
```

Comment:
Shall we just call it `ModuleTracker`? It's more intuitive than `ModTracker`

Reply:
There already exists a ``ModuleTracker`` in torch.utils.ModuleTracker. It has some bugs and I needed more functionality than it currently provides. So I had submitted a PR #127624, but Alban is busy and needs more time to review. So Gokul suggested I land my own version in distributed, hence calling it ``ModTracker``. Open for other name suggestions as well.